File: scripts/signature_wrapper.py
import hmac
import hashlib
import time
import secrets
import json

class A2ASignatureWrapper:
    def __init__(self, agent_id, secret):
        self.agent_id = agent_id
        self.secret = secret

    def wrap(self, message_type, data, resource_pointers=None, conatus_weight=None):
        header = {
            "agent_id": self.agent_id,
            "alg": "HS256",
            "nonce": secrets.token_hex(16),
            "timestamp": int(time.time() * 1000)
        }
        
        if resource_pointers:
            header["resource_pointers"] = resource_pointers

        payload = {
            "type": message_type,
            "data": data
        }
        
        if conatus_weight is not None:
            payload["conatus_weight"] = conatus_weight

        # Sign the packet
        message = json.dumps({"header": header, "payload": payload}, sort_keys=True)
        signature = hmac.new(
            self.secret.encode(),
            message.encode(),
            hashlib.sha256
        ).hexdigest()

        return {
            "header": header,
            "payload": payload,
            "signature": signature
        }

File: packages/a2a-core/src/signature-wrapper.ts
import { getRandomBytes, hmacSha256 } from '@the-new-fuse/security';

/**
 * TNF Resource Pointer (TRP)
 * Represents a reference to external data stored in the Federated Knowledge Graph
 * or another persistent store (e.g. pgvector, S3, R2, local filesystem).
 */
export interface TNFResourcePointer {
  uri: string; // e.g. pgvector://shards/shard-123, file:///...
  integrityHash?: string;
  mimeType?: string;
  size?: number;
}

/**
 * A2A Signed Packet (DACC-v1 Compliant)
 *
 * Evolution: Now supports resource_pointers to prevent "All-in-Memory" OOM crashes.
 * Payload should contain lightweight metadata, while heavy data is passed via pointers.
 */
export interface A2ASignedPacket {
  header: {
    agent_id: string;
    alg: 'HS256';
    nonce: string;
    timestamp: number;
    resource_pointers?: Record<string, TNFResourcePointer>;
  };
  payload: {
    type: string;
    channel?: string;
    data: unknown;
    conatus_weight?: number;
  };
  signature: string;
}

export class A2ASignatureWrapper {
  constructor(
    private readonly agentId: string,
    private readonly secret: string
  ) {}

  /**
   * Wraps an A2A message with a DACC-v1 compliant signature and optional resource pointers.
   */
  wrap(
    type: string,
    data: unknown,
    options?: {
      channel?: string;
      conatusWeight?: number;
      resourcePointers?: Record<string, TNFResourcePointer>;
    }
  ): A2ASignedPacket {
    const header: A2ASignedPacket['header'] = {
      agent_id: this.agentId,
      alg: 'HS256' as const,
      nonce: this.generateNonce(),
      timestamp: Date.now(),
      resource_pointers: options?.resourcePointers,
    };

    const payload = {
      type,
      channel: options?.channel,
      data,
      conatus_weight: options?.conatusWeight,
    };

    const message = JSON.stringify({ header, payload });
    const signature = hmacSha256(message, this.secret);

    return { header, payload, signature };
  }

  private generateNonce(): string {
    return getRandomBytes(16).toString('hex');
  }
}
